h2/utilities.py

```diff
@@ -214,6 +217,22 @@ def _reject_uppercase_header_fields(headers, hdr_validation_flags):
         yield header
 
 
+def _reject_surrounding_whitespace(headers, hdr_validation_flags):
+    """
+    Raises a ProtocolError if any header name or value is surrounded by
+    whitespace characters.
+    """
+    for header in headers:
+        if header[0].strip() != header[0]:
+            raise ProtocolError(
+                "Received header name surrounded by whitespace %s" % header[0])
+        if header[1].strip() != header[1]:
+            raise ProtocolError(
+                "Received header value surrounded by whitespace %s" % header[1]
+            )
+        yield header
```

Comment:
So for a first pass I think this is totally fine. If you want an extension task, it would be really interesting to prototype different ways we could do this along with benchmarks, to see how they perform. This code is in a quite hot section of the code, which means that any efficiencies we can eke out of each function are huge.

This doesn't block the PR at all, just a suggestion for extension work. :grin: